Generics error.

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relevant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
**V version:** V 0.3.1 721328e
**OS:** macOS 12.4

<!-- Please include the standalone minimal reproducible code.
If not, it may be closed by the administrators. -->
**What did you do?**

```v
import vweb

struct ApiSuccessResponse<T> {
	success bool
	result  T
}

struct App {
	vweb.Context
}

fn (mut app App) json_success<T>(result T) vweb.Result {
	response := ApiSuccessResponse<T>{
		success: true
		result: result
	}

	return app.json(response)
}

fn main() {
	mut app := &App{}

	vweb.run(app, 80)
}

```

**What did you expect to see?**
Successful compilation.
 
**What did you see instead?**
```
==================
   ^~~~~~~~~~~~~~~
/tmp/v_501/generic.13395306593623544363.tmp.c:29838:8: error: implicit declaration of function 'main__App_json_success' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
                                                        main__App_json_success(app, ((string*)args.data) [0] );
                                                        ^
/tmp/v_501/generic.13395306593623544363.tmp.c:29889:7: error: implicit declaration of function 'main__App_json_success' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
                                                main__App_json_success(app, ((string*)method_args.data) [0] );
                                                ^
/tmp/v_501/generic.13395306593623544363.tmp.c:29918:4: warning: expression result unused [-Wunused-value]
  (*(int*)_t22.data);
   ^~~~~~~~~~~~~~~~
/tmp/v_501/generic.13395306593623544363.tmp.c:30112:4: warning: expression result unused [-Wunused-value]
  (*(int*)_t1.data);
...
```

--- a/vlib/v/gen/c/comptime.v
+++ b/vlib/v/gen/c/comptime.v
@@ -467,12 +467,18 @@ fn (mut g Gen) comptime_for(node ast.ComptimeFor) {
 		if methods.len > 0 {
 			g.writeln('FunctionData $node.val_var = {0};')
 		}
-		for method in methods { // sym.methods {
-			/*
-			if method.return_type != vweb_result_type { // ast.void_type {
-				continue
+		for method in methods {
+			// filter vweb route methods (non-generic method)
+			if method.receiver_type != 0 {
+				rec_sym := g.table.sym(method.receiver_type)
+				if rec_sym.kind == .struct_ {
+					if _ := g.table.find_field_with_embeds(rec_sym, 'Context') {
+						if method.generic_names.len > 0 {
+							continue
+						}
+					}
+				}
 			}
-			*/
 			g.comptime_for_method = method.name
 			g.writeln('/* method $i */ {')
 			g.writeln('\t${node.val_var}.name = _SLIT("$method.name");')

--- a/vlib/v/gen/c/comptime.v
+++ b/vlib/v/gen/c/comptime.v
@@ -467,9 +467,10 @@ fn (mut g Gen) comptime_for(node ast.ComptimeFor) {
 		if methods.len > 0 {
 			g.writeln('FunctionData $node.val_var = {0};')
 		}
+		typ_vweb_result := g.table.find_type_idx('vweb.Result')
 		for method in methods {
 			// filter vweb route methods (non-generic method)
-			if method.receiver_type != 0 {
+			if method.receiver_type != 0 && method.return_type == typ_vweb_result {
 				rec_sym := g.table.sym(method.receiver_type)
 				if rec_sym.kind == .struct_ {
 					if _ := g.table.find_field_with_embeds(rec_sym, 'Context') {

--- a/vlib/vweb/vweb_app_test.v
+++ b/vlib/vweb/vweb_app_test.v
@@ -64,3 +64,17 @@ fn (mut app App) time_json_pretty() {
 		'time': time.now().format()
 	})
 }
+
+struct ApiSuccessResponse<T> {
+	success bool
+	result  T
+}
+
+fn (mut app App) json_success<T>(result T) vweb.Result {
+	response := ApiSuccessResponse<T>{
+		success: true
+		result: result
+	}
+
+	return app.json(response)
+}

--- a/vlib/vweb/vweb_app_test.v
+++ b/vlib/vweb/vweb_app_test.v
@@ -78,3 +78,17 @@ fn (mut app App) json_success<T>(result T) vweb.Result {
 
 	return app.json(response)
 }
+
+// should compile, this is a helper method, not exposed as a route
+fn (mut app App) some_helper<T>(result T) ApiSuccessResponse<T> {
+	response := ApiSuccessResponse<T>{
+		success: true
+		result: result
+	}
+	return response
+}
+
+// should compile, the route method itself is not generic
+fn (mut app App) ok() vweb.Result {
+	return app.json(app.some_helper(123))
+}

